apps/bfd-server-ng/src/main/java/gov/cms/bfd/server/ng/eob/EobHandler.java

```diff
@@ -71,18 +103,90 @@ public Bundle searchByBene(
+   * @param samhsaFilterMode SAMHSA filter mode
    * @return bundle
    */
   public Bundle searchById(
-      Long claimUniqueId, DateTimeRange serviceDate, DateTimeRange lastUpdated) {
-    var eob = searchByIdInner(claimUniqueId, serviceDate, lastUpdated);
+      Long claimUniqueId,
+      DateTimeRange serviceDate,
+      DateTimeRange lastUpdated,
+      SamhsaFilterMode samhsaFilterMode) {
+    var eob = searchByIdInner(claimUniqueId, serviceDate, lastUpdated, samhsaFilterMode);
     return FhirUtil.bundleOrDefault(eob.map(e -> e), claimRepository::claimLastUpdated);
   }
 
   private Optional<ExplanationOfBenefit> searchByIdInner(
-      Long claimUniqueId, DateTimeRange serviceDate, DateTimeRange lastUpdated) {
-    var claim = claimRepository.findById(claimUniqueId, serviceDate, lastUpdated);
+      Long claimUniqueId,
+      DateTimeRange serviceDate,
+      DateTimeRange lastUpdated,
+      SamhsaFilterMode samhsaFilterMode) {
+    var claimOpt = claimRepository.findById(claimUniqueId, serviceDate, lastUpdated);
+    if (claimOpt.isEmpty()) {
+      return Optional.empty();
+    }
+    var claim = claimOpt.get();
+
+    if (samhsaFilterMode == SamhsaFilterMode.EXCLUDE && claimHasSamhsa(claim)) {
+      return Optional.empty();
+    }
+    return Optional.of(claim.toFhir());
+  }
+
+  private boolean isCodeSamhsa(String targetCode, LocalDate claimDate, SecurityLabel entry) {
+    return isClaimDateWithinBounds(claimDate, entry) && entry.matches(targetCode);
+  }
+
+  // Returns true if the given claim contains any procedure that matches a SAMHSA
+  // security label code from the dictionary.
+  private boolean claimHasSamhsa(Claim claim) {
+    var claimThroughDate = claim.getBillablePeriod().getClaimThroughDate();
+    var drgSamhsa = drgCodeIsSamhsa(claim, claimThroughDate);
+    var claimItemSamhsa =
+        claim.getClaimItems().stream().anyMatch(e -> claimItemIsSamhsa(e, claimThroughDate));
+
+    return drgSamhsa || claimItemSamhsa;
+  }
+
+  private boolean claimItemIsSamhsa(ClaimItem claimItem, LocalDate claimThroughDate) {
+    return procedureMatchesSamhsaCode(claimItem.getClaimProcedure(), claimThroughDate)
+        || hcpcsCodeMatches(claimItem.getClaimLine(), claimThroughDate);
+  }
+
+  private boolean drgCodeIsSamhsa(Claim claim, LocalDate claimDate) {
+    var entries = SECURITY_LABELS_MAP.get(SystemUrls.CMS_MS_DRG);
+    var drg = claim.getDrgCode().map(Object::toString).orElse("");
+    return entries.stream().anyMatch(e -> isCodeSamhsa(drg, claimDate, e));
+  }
+
+  private boolean hcpcsCodeMatches(ClaimLine claimLine, LocalDate claimDate) {
+    var hcpcs = claimLine.getHcpcsCode().getHcpcsCode().orElse("");
+    return Stream.of(SystemUrls.AMA_CPT, SystemUrls.CMS_HCPCS)
+        .flatMap(s -> SECURITY_LABELS_MAP.get(s).stream())
+        .anyMatch(c -> isCodeSamhsa(hcpcs, claimDate, c));
+  }
+
+  // Checks ICDs.
+  private boolean procedureMatchesSamhsaCode(ClaimProcedure proc, LocalDate claimDate) {
```

Comment:
feels a little weird to name this `procedureMatchesSamhsaCode` then check procedure and diagnosis codes